vassal-app/src/main/java/VASSAL/build/module/GameRefresher.java

```diff
@@ -689,25 +618,50 @@ private GamePiece processGamePiece(GamePiece piece, Command command) {
     return newPiece;
   }
 
-  private Point getDeckFreePosition(Deck deck) {
-    Point tempPosition = new Point(-1, -1);
-    Boolean correctTempPositionNotFound;
-    final GamePiece[] pieces;
-    pieces = deck.getMap().getAllPieces();
-    do {
-      correctTempPositionNotFound = false; //Assuming scan of pieces finds no match on tempPosition
-      for (final GamePiece piece : pieces) {
-        final Point piecePosition = piece.getPosition();
-        if (piecePosition.equals(tempPosition)) {
-          tempPosition.x -= 1;
-          correctTempPositionNotFound = true;
-          break;
-        }
-      }
-    } while (correctTempPositionNotFound);
-    return tempPosition;
-  }
+  /**
+   * Process the pieces in a Deck
+   * @param deck Deck to procss
+   * @param command Generated Command
+   */
+  private void processDeck(Deck deck, Command command) {
 
+    // Take a copy of the pieces in the Deck
+    final List<GamePiece> pieces = deck.asList();
+
+    // Remove all the pieces from the Deck by moving them to 0,0 then deleting them. This ensures the Undo
+    // ends with a valid MovePiece command to move the piece back into the Deck
+    for (final GamePiece piece : pieces) {
+      final Command move = deck.getMap().placeAt(piece, new Point(0, 0));
```

Comment:
I guess 0,0 is safe because we're just moving the piece there, not processing it into stacks, decks, etc? (I mean I realize -3000,-3000 wouldn't be safe either, haha, but just checking that it doesn't matter if anything else if "already there" at 0,0)

Reply:
map.placeAt() will place the piece there as a non-stacking piece before it is immediately deleted. It ensures the final Command of the undo sequence is a MovePiece moving the piece back into the Deck which is the only way I could find to cleanly undo the refresh. Deleting the piece directly from the Deck leaves the final Undo Command as an AddPiece which does not set the piece correctly into the Deck. Which makes me wonder if there is a bug if you delete cards directly from a Deck using a GKC that they will not correctly Undo?